File: data_structures/go/linked_list.go
package ds

import "fmt"

type LLNode struct {
	Next *LLNode
	Prev *LLNode
	Val  interface{}
}

type LinkedList struct {
	Head *LLNode
	Tail *LLNode
	size int
}

func (ll *LinkedList) AddHead(x interface{}) {
	// Wrap the data in a node and add to head.
	node := LLNode{Val: x}
	if ll.size == 0 {
		ll.addFirstNode(&node)
	} else {
		node.Next = ll.Head
		ll.Head.Prev = &node
		ll.Head = &node
	}
	ll.size++
}

func (ll *LinkedList) AddTail(x interface{}) {
	node := LLNode{Val: x}
	if ll.size == 0 {
		ll.addFirstNode(&node)
	} else {
		node.Prev = ll.Tail
		ll.Tail.Next = &node
		ll.Tail = &node
	}
	ll.size++
}

func (ll *LinkedList) addFirstNode(n *LLNode) {
	ll.Head = n
	ll.Tail = n
}

func (ll *LinkedList) removeLastNode() interface{} {
	x := ll.Head.Val
	ll.Head.Next = nil
	ll.Head.Prev = nil
	ll.Head = nil
	ll.Tail = nil
	return x
}

func (ll *LinkedList) PopHead() interface{} {
	ll.size--

	if ll.size == 0 {
		return ll.removeLastNode()
	}

	x := ll.Head.Val
	oldNode := ll.Head
	oldNode.Next.Prev = nil
	ll.Head = oldNode.Next
	oldNode.Next = nil

	return x
}

func (ll *LinkedList) PopTail() interface{} {
	ll.size--

	if ll.size == 0 {
		return ll.removeLastNode()
	}

	x := ll.Tail.Val
	oldNode := ll.Tail
	oldNode.Prev.Next = nil
	ll.Tail = oldNode.Prev
	oldNode.Prev = nil

	return x
}

func (ll *LinkedList) PeekHead() interface{} {
	return ll.Head.Val
}

func (ll *LinkedList) PeekTail() interface{} {
	return ll.Tail.Val
}

func (ll *LinkedList) Size() interface{} {
	return ll.size
}

func (ll *LinkedList) IsEmpty() bool {
	return ll.Size() == 0
}

func (ll *LinkedList) Display() {
	fmt.Printf("[")
	node := ll.Head
	for node != nil {
		fmt.Printf("%v", node.Val)
		if node.Next != nil {
			fmt.Printf(" -> ")
		}
		node = node.Next
	}
	fmt.Printf("]\n")
}

File: data_structures/python/linked_list.py
class Node:
    def __init__(self, x):
        self.next: 'Node' = None
        self.prev: 'Node' = None
        self.x = x


class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self._size: int = 0

    # ===================================================================================================
    # Add an element x (can be any type) to the head or tail.
    # ===================================================================================================

    def add_head(self, x):
        node = Node(x)
        if self.size() == 0:
            self.head = node
            self.tail = node
        else:
            self.head.prev = node
            node.next = self.head
            self.head = node
        self._size += 1

    def add_tail(self, x):
        node = Node(x)
        if self.size() == 0:
            self.head = node
            self.tail = node
        else:
            self.tail.next = node
            node.prev = self.tail
            self.tail = node
        self._size += 1

    # ===================================================================================================
    # Pop an element from the head or tail, and return it.
    # ===================================================================================================

    def pop_head(self):
        x, result = self._final_removal()
        if result:
            return x
        else:
            x = self.head.x
            next_node = self.head.next
            next_node.prev = None
            self.head.next = None
            self.head = next_node
            return x

    def pop_tail(self):
        x, result = self._final_removal()
        if result:
            return x
        else:
            x = self.tail.x
            prev_node = self.tail.prev
            prev_node.next = None
            self.tail.prev = None
            self.tail = prev_node
            return x

    def _final_removal(self):
        if self.size() == 0:
            return None, True
        self._size -= 1
        if self.size() == 0:
            x = self.head.x
            self.head = None
            self.tail = None
            return x, True
        else:
            return None, False

    # ===================================================================================================
    # Return the head or tail element without removing it.
    # ===================================================================================================

    def peek_head(self):
        if self.head is not None:
            return self.head.x
        else:
            return None

    def peek_tail(self):
        if self.tail is not None:
            return self.tail.x
        else:
            return None

    # ===================================================================================================
    # Return the size of the list.
    # ===================================================================================================

    def size(self) -> int:
        return self._size

    # ===================================================================================================
    # Optional Utility Functions
    # ===================================================================================================

    def __repr__(self):
        elements = []
        node = self.head
        while node is not None:
            elements.append(node.x)
            node = node.next
        content = " -> ".join(map(str, elements))
        return f"[{content}]"

    def display(self):
        print(self)
